### ISSUE TITLE
spi boot usb hard drive

### ISSUE THREAD

**Issue by kfox1111** [ayufan-rock64/linux-build#318]
I'm trying to boot from usb hard drive via the newest spi flashed image on a rock64 cluster. It is working about half the time. Sometimes it detects the drive, sometimes it says it finds a device but no drive.

I'm guessing it is taking a bit too long for the drive to spin up and be detected by the usb adapter. Is there any way to delay a bit longer before drive probing?

**Comment by XDA-Bam** [comment]
The spin-up time is likely not the problem. See issue https://github.com/ayufan-rock64/linux-build/issues/257 . To the best of my knowledge, there is currently no workaround for this issue. If you find one, please let me know.

**Comment by LipuFei** [comment]
Same issue for me. I have this problem with the USB3 port. After I switched my HDD to a USB2 port, it seems to work fine.

**Comment by Wrong-Code** [comment]
Add me to the list. I have a SSD drive connected to the USB3 port, SPI booted it normally on the first days, and then progressively it failed to boot half of the times until yesterday, where it stopped completely. After many many tries, I booted from the SD card and by mounting manually the SSD partition I could check the partition was in clean state (so there is still more than enough current to drive the disk), yet the Rock64 SBC continued in refusing to boot from it.

After having read about booting from the USB2 port, I tried on my Rock64 and it booted effectively once I hooked the SSD there. Not content, I powered off the card and move the SSD again on the USB3 port, and this time it worked. There is definitively something wrong in the boot code (or in the hardware) regarding the USB3 port.

**Comment by henricazottes** [comment]
Yup, similar problem here: USB3 works with usb  flash drive but not with the ssd. My adapter has two cables (one for data and one for power) and plugging it to an external power supply doesn't help.

Any update on that ?

**Comment by XDA-Bam** [comment]
This seems to be fixed now, for me. Recently updated to U-Boot revision [1063](https://github.com/ayufan-rock64/linux-u-boot/releases/tag/2017.09-rockchip-ayufan-1063-g29843fbd42) and have not encountered the problem anymore since then on my Rock64. Updated to the most recent [1065](https://github.com/ayufan-rock64/linux-u-boot/releases/tag/2017.09-rockchip-ayufan-1065-g95f6152134) five days ago and just now tested it again - three restarts in a row without hiccups. Booting fine, directly from a USB3-attached SSD (no SD card, no EMMC).

@henricazottes @kfox1111 Can you confirm this on your devices?

**Comment by kfox1111** [comment]
Thanks for the info. :)
I'll try and verify next chance I get.

**Comment by Wrong-Code** [comment]
@XDA-Bam Today I've had some time to check U-Boot version 1065 on my Rock64 V2 running Armbian Bionic 5.90. Unfortunately the issue persists, I cannot boot from USB3 (but I can still boot from USB2)

Methodology:

- grabbed the relevant .deb file
- installed with apt install -f debfile
- ran rock64_write_spi_flash.sh to update the SPI

If I am missing something here please tell me. Many thanks for the effort.

**Comment by XDA-Bam** [comment]
@Wrong-Code I'm running Armbian Buster 5.91 with U-Boot 1065, but my hardware is Rock64 v1 as far as I can tell. I etched the fresh Armbian image directly to my SSD and updated/installed everything from there. U-Boot flash was done using the SD-card method with LAN and SSD disconnected.

Hard to tell from the distance what may cause the difference in outcomes for you and me. It might be, that relevant U-Boot patches that made it work for Rock64 v1 did not apply to or did'n fix it for v2. I would first check if you are really running U-Boot 1065 right now. If so, there's probably nothing you can do except to wait for a fix :-/

**Comment by Wrong-Code** [comment]
@XDA-Bam I use Armbian Bionic 5.98. Here is the U-Boot version:

```
# strings /dev/mtd0 | grep U-Boot
U-Boot SPL 2017.09-rockchip-ayufan-1065-g95f6152134 (Aug 26 2019 - 12:40:32)

```
I am afraid it could be due to either something in V2 boards, or to the SATA-USB adapter I am using:

```
# lsusb
ID 152d:0578 JMicron Technology Corp. / JMicron USA Technology Corp. 

# lsusb -t
/:  Bus 05.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/1p, 5000M
    |__ Port 1: Dev 2, If 0, Class=Mass Storage, Driver=uas, 5000M

```

Another possibility would be hook a TTY to Rock64 and see if the u-boot messages give some hint, though I doubt it. I'd have to open my Pine metal box and to found some fresh thermal grease to reseal the box afterwards.

**Comment by XDA-Bam** [comment]
OK, so our U-Boot is identical as expected. My USB-SATA bridge has a different PID, though:
`
ID 152d:1576 JMicron Technology Corp. / JMicron USA Technology Corp.

/:  Bus 05.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/1p, 5000M
    |__ Port 1: Dev 2, If 0, Class=Mass Storage, Driver=uas, 5000M
`
Yours seems to be a [JMicron JMS567](https://usb-ids.gowdy.us/read/UD/152d/0578) ([datasheet](http://www.jmicron.com/PDF/brief/jms567.pdf)). Mine is probably a JMS576 ([datasheet](http://www.jmicron.com/PDF/brief/jms576.pdf)). Specs look _very_ similar to me.

**Comment by Wrong-Code** [comment]
Yours is a newer model than mine, apparently, supporting also USB 3.1

The curious thing is that I seem to remember to have this controller booting 2-3 times out of 10 one year ago, or a little less than that. Since then it fails constantly to boot from USB3.

**Comment by luckcolors** [comment]
It seems to be a problem with this JMicron chip in particular.
Just used anothe radaptor with VIA Labs, Inc. and it booted.

**Comment by D4rkSl4ve** [comment]
Can Rock64 boot from USB3(blue) connector?  Trying to get an SSD drive as the main drive (boot/docker containers).

**Comment by kskewes** [comment]
> Can Rock64 boot from USB3(blue) connector? Trying to get an SSD drive as the main drive (boot/docker containers).

It seems very hit and miss for me. Sometimes on first power up but never reliable enough to run OS off it.
Using sd or emmc card for boot and then usb3 secondary seems fine.

**Comment by kfox1111** [comment]
Yeah. I couldn't ever get them to reliably boot from usb3 and that is why I bought them in the first place. I had to switch away from the hardware unfortunately.

**Comment by Wrong-Code** [comment]
I have bought in the mean time a couple of USB 3 connected storage cases, with different UAS chip and all showing regrettably the same issues.

We are not alone, however: I have also bought in the mean time a Raspberry Pi 4, think about switching to a better supported platform, and while their latest firmware can boot using two out of three of my USB 3 cases, reliability is a completely different matter, either with Raspbian, CentOS or Ubuntu. I come to conclusion that USB 3 in the ARM SBC world is something really underdeveloped, and unfortunately underpowered too. There are suggestions about adding a reliable USB 3 powered hub to deal with the issue, which has the drawback of adding more cases and cables to the mix.

It's a pity, but I will not invest any more money on this stuff.
